src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubReceiver.java

```diff
@@ -0,0 +1,506 @@
+        if (ae.getStatusCode().getCode().equals(StatusCode.Code.ALREADY_EXISTS)) {
+          return;
+        }
+
+        //This error is thrown is the Topic does not exist.
+        // Call the stop method so the pipeline fails.
+        if (ae.getStatusCode().getCode().equals(StatusCode.Code.NOT_FOUND)) {
+          String message = String.format(MISSING_TOPIC_ERROR_MSG, topic, project);
+          stop(message, ae);
+          return;
+        }
+
+        //Retry if the exception is retriable.
+        if (ae.isRetryable()) {
+          backoff = sleepAndIncreaseBackoff(backoff);
+          continue;
+        }
+
+        //Report that we were not able to create the subscription and stop the receiver.
+        stop(String.format(CREATE_SUBSCRIPTION_ERROR_MSG, subscription), ae);
+        return;
+      } catch (IOException ioe) {
+        //Report that we were not able to create the subscription admin client and stop the receiver.
+        stop(String.format(CREATE_SUBSCRIPTION_ADMIN_CLIENT_ERROR_MSG, subscription), ioe);
+        return;
+      }
+    }
+
+    //If we were not able to create the subscription after 5 attempts, stop the pipeline and report the error.
+    stop(String.format(CREATE_SUBSCRIPTION_RETRY_ERROR_MSG, subscription), lastApiException);
+  }
+
+  /**
+   * Build subscriber client and schedule fetch task to run at a 100 millisecond delay.
+   *
+   * @throws ApiException     when the Pub/Sub API throws a non-retryable exception.
+   * @throws RuntimeException when the GrpcSubscriberStub cannot be created.
+   */
+  public void scheduleFetch() {
+    try {
+      SubscriberStubSettings subscriberSettings = getSubscriberSettings();
+      subscriber = getSubscriber(subscriberSettings);
+
+      executor.scheduleWithFixedDelay(this::receiveMessages, 0, 100, TimeUnit.MILLISECONDS);
+    } catch (IOException ioe) {
+      //This exception is thrown when the subscriber could not be created.
+      //Report the exception and stop the receiver.
+      String message =
+        String.format(SUBSCRIBER_ERROR_MSG, subscription, project);
+      stop(message, ioe);
+    }
+  }
+
+  /**
+   * Fetch new messages for our subscription.
+   * Implements exponential backoff strategy when a retryable exception is received.
+   * This method stops the receiver if a non retryable ApiException is thrown by the Google Cloud subscriber client.
+   */
+  protected void receiveMessages() {
+    int backoff = backoffConfig.getInitialBackoffMs();
+
+    //Try with backoff until stopped or the task succeeds.
+    while (!isStopped()) {
+      try {
+        fetchAndAck();
+        return;
+      } catch (ApiException ae) {
+        if (ae.isRetryable()) {
+          backoff = sleepAndIncreaseBackoff(backoff);
+        } else {
+          //Restart the receiver if the exception is not retryable.
+          String message =
+            String.format(FETCH_ERROR_MSG, subscription, project);
+          restart(message, ae);
```

Comment:
what's the difference between restarting the receiver and retrying the message fetch?

Reply:
A retryable exception is something like "429 too many requests", which means the request can be safely retried at a later time (in this case, the exponential backoff logic starts increasing the delay for requests).

A non-retryable exception can be something related to credentials expired, subscription deleted, or an internal server error from Pub/Sub. In this case, the receiver gets restarted, credentials re-initialized, subscription recreated if needed, and then messages start pulling once again.

Reply:
I see. Can you add a comment with this info?

Reply:
I have added this comment.